### PR TITLE
[inductor] Avoid bool being upcast to int

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109913

Currently the inductor code for `x.any(-1)` does a this strange dance:
```python
tmp0 = tl.load(in_ptr0 + (r1 + (128*x0)), rmask & xmask)
tmp1 = tmp0.to(tl.int64)
tmp2 = (tmp1 != 0)
```

This happens because `register_lowering` is doing type promotion with the
dimension argument, and so promotes to `int64` which we then cast back to bool.
A better fix would be to fix `register_lowering` but for now I just remove
the unnecessary type promotion from `aten.any`.

In the current code we also see:
```python
     tmp5 = tl.where(rmask & xmask, tmp3, 0)
```
which promotes the boolean value to int since `0` is an int32 in triton.
This fixes it to generate a boolean constant instead.

Finally there is also a triton bug where the `tl.load` itself upcasts to
`tl.int8`. I fix this by adding an explicit cast to `tl.int1`. The final
kernel code looks like:

```python
tmp0 = tl.load(in_ptr0 + (r1 + (128*x0)), rmask & xmask).to(tl.int1)
tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
tmp3 = tl.full([1, 1], 0, tl.int1)
tmp4 = tl.where(rmask & xmask, tmp1, tmp3)
tmp5 = triton_helpers.any(tmp4, 1)[:, None]

```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1544,8 +1549,13 @@ def final_argreduce(buffer, result_var, value, index):
             default = self._map_tuple_or_scalar(triton_constant, default)
 
             def _mask_value(value, default):
+                ttype = triton_compute_type(src_dtype)
+                other = self.cse.generate(
+                    self.compute,
+                    f"tl.full({[1] * self.triton_tensor_ndim()}, {default}, {ttype})",
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1555]: `tl.where` does not broadcast on the number of dims?

**Review by lezcano — APPROVED** [review]

**Comment by malfet** [comment]
> @pytorchbot revert -m "causing performance regression in relevant metrics, @malfet I believe you are the correct person to help identify and fix the issues. More details check internal OPS count for ads metricsnin the internal related diff" -c nosignal

Please note, that revert category should be `ghfirst`

**Comment by peterbell10** [comment]
@malfet are you able to share any details of the test failure?

**Comment by peterbell10** [comment]
@malfet ping

**Pull request closed without merging**